src/HomologicalAlgebra/BoundedCocomplexes.jl

```diff
@@ -0,0 +1,397 @@
+#     category. Unfortunately, we can not derive this from a type 
+#     directly, but we must provide another concrete instance of 
+#     an object in the same category, i.e.
+#
+#       zero_object(::GroupType)::GroupType
+#
+#     must be implemented for every `GroupType` that is being used.
+#   * For any concrete instances of `GroupType1` and `GroupType2` 
+#     there must be a constructor for the zero morphism:
+#
+#       zero_morphism(::GroupType1, ::GroupType2)
+#
+#   * For a homomorphism f : C → D with C of type `GroupType1` 
+#     and D of type `GroupType2` there must be the functionality 
+#
+#       kernel(f)::KernelType
+#       image(f)::ImageType
+#
+#     returning the kernel and the image as subgroups of C and D, 
+#     respectively.
+#   * For a group Z and a subgroup B ⊂ Z there must be the function
+#
+#       quo(Z, B)::QuotientType
+#
+#     returning the quotient Z/B.
+#
+#   * There must be direct sums a vector of groups:
+#
+#       direct_sum(::Vector{GroupType})
+#
+#     for all types of groups that are being used. 
+#     For a vector c = (C₁,…,Cₙ) this must return a triple 
+#
+#       (C₁ ⊕ … ⊕ Cₙ, (i₁,…,iₙ), (p₁,…, pₙ))
+#
+#     consisting of the direct sum C₁ ⊕ … ⊕ Cₙ, the injections 
+#
+#       iₖ : Cₖ ↪ C₁ ⊕ … ⊕ Cₙ
+#
+#     and the projections 
+#
+#       pₖ : C₁ ⊕ … ⊕ Cₙ → Cₖ.
+#
+#   * There must be the Hom-functor for pairs of groups 
+#
+#       Hom(::GroupType1, ::GroupType2)::GroupType3
+#
+#     taking a pair (C, D) to Hom(C, D).
+#
+########################################################################
+
+import Base: last, length, first
+export ChainCocomplex, BoundedCocomplex
+export differential, differentials, shift, kernels, cochains, boundaries, cohomologies, kernel, boundary, cohomology
+
+export ChainCocomplexHom, BoundedCocomplexHom
+export domain, codomain, maps, shift, getindex, mapping_cone
+
+abstract type ChainCocomplex{GroupType, GroupHomType} end
+
+########################################################################
+# 
+# Bounded cochain complexes of Abelian groups.
+#
+# These store a finite number of Abelian groups Xⁱ together with 
+# morphisms dⁱ: Xⁱ→ Xⁱ⁺¹ and fills up all remaining entries with 
+# zeroes. Example:
+#
+# math. index:   -5   -4   -3   -2   -1    0    1    2    3
+#
+# groups:         0    X⁻⁴→ X⁻³→ X⁻²→ X⁻¹→ X⁰→  X¹   0    0
+#
+# intern. index:  -    1    2    3    4    5    6    -    - 
+# 
+# Internally, the groups and morphisms are stored in vectors 
+# starting with the `internal index` 1. For the translation from 
+# the mathematical to the internal index, we also store a shift, 
+# which is 5 in the above case.
+#
+########################################################################
+
+mutable struct BoundedCocomplex{GroupType, GroupHomType} <: ChainCocomplex{GroupType, GroupHomType}
+  shift::Int
+  C::Vector{GroupType}
+  d::Vector{GroupHomType}
+
+  # fiends for caching
```

Comment:
That sounds quite devilish :-)
```suggestion
  # fields for caching
```